### PR TITLE
feat: keyboard simulation

### PR DESCRIPTION
**Pull request by hatayama**
## Summary
- Add simulate-keyboard MCP tool for injecting key input via Input System in PlayMode
- Add UnityChan assets with a new Input System WASD controller (8-directional movement, shift-to-walk, jump)
- Set up Cinemachine third-person follow camera
- Register all 20 unused animations into the AnimatorController
- Create automated demo script (scripts/unitychan-demo.sh) for showcasing motions

### simulate-keyboard MCP Tool
- Supports Press (tap), KeyDown (hold), and KeyUp (release) actions
- Injects key input directly into PlayMode Input System
- Includes keyboard demo scene with visual key feedback

### UnityChan Setup
- Add UnityChan model, animations, and materials
- Add missing packages: animation, toonshader, unityanalytics, imageconversion, cinemachine
- UnityChanWASDController.cs: new Input System, 8-directional movement, shift walk, jump from idle/locomotion
- Register all 20 unused animation clips as states in UnityChanLocomotions controller
- Cinemachine virtual camera with smooth third-person follow

### Demo Script
- scripts/unitychan-demo.sh: starts PlayMode, runs walk/run/jump choreography, stops PlayMode
- Circular movement patterns to stay near center of ground plane
- hold_for helper for safe key release, trap for cleanup on interruption

## Test Plan
- [ ] Play SimulateKeyboardDemoScene and verify keyboard simulation works
- [ ] Verify UnityChan moves with WASD/arrow keys, walks with Shift, jumps with Space
- [ ] Run sh scripts/unitychan-demo.sh and verify the demo completes without falling off
- [ ] Verify Cinemachine camera follows the character smoothly

## Overview

This PR adds PlayMode keyboard-input simulation tooling, an editor utility to build a keyboard demo scene, PlayMode tests for the simulation tool, demo UI components, editor helpers for UnityChan, and skill documentation for the uloop simulate-keyboard capability. It also tightens an existing simulate-mouse test timeout and removes an obsolete simulate-mouse SKILL doc.

### Simulate-Keyboard MCP Tool (behavior & tests)
- Comprehensive PlayMode test suite: Assets/Tests/PlayMode/SimulateKeyboardTests.cs
  - Tests Press (tap), Press with duration, KeyDown/KeyUp lifecycles, multi-key holds, ReleaseAllKeys, invalid/empty key handling, and verifies wasPressedThisFrame behavior (FramePressObserver).
  - Adds RunTool helper with bounded wait for tool execution.
- Tests exercise direct Input System injection semantics and hold-state management.

(Note: the main tool implementation source file is described in PR notes, but this summary is limited to files present in this change set; tests exercise the tool API.)

### Demo UI and Scene Builder
- DemoKeyIndicator: runtime UI component showing key press feedback (Assets/Tests/Demo/DemoKeyIndicator.cs).
  - Configurable target Key and label; toggles color and slight scale on press.
- DemoKeyboardSceneBuilder: Editor script to programmatically build/saving the keyboard demo scene (Assets/Tests/Demo/Editor/DemoKeyboardSceneBuilder.cs).
  - Creates Camera, EventSystem with InputSystemUIInputModule, Canvas, title/status, and full keyboard layout UI; exposed via a Unity MenuItem.
- Scene asset present: Assets/Scenes/SimulateKeyboardDemoScene.unity (scene saved by builder).

### UnityChan Editor Helper
- CreateLocatorHere: small Editor menu/context utility for creating locator GameObjects under a selected transform (Assets/UnityChan/Editor/CreateLocatorHere.cs).

### Skill Documentation
- Adds .claude/skills/uloop-simulate-keyboard/SKILL.md describing the simulate-keyboard CLI/skill (usage, parameters, examples).
- Removes obsolete .claude/skills/simulate-mouse-demo/SKILL.md.

### Small Behavioral Change
- Assets/Tests/PlayMode/SimulateMouseTests.cs: adds a 5-second timeout when awaiting asynchronous tool execution to avoid indefinite hangs.

## Package & Asset Notes (from PR description)
- PR references adding packages and UnityChan assets (animation, toonshader, unityanalytics, imageconversion, cinemachine) and a UnityChan WASD controller and Animator changes. In the current repository snapshot:
  - The UnityChan WASD controller source and an Animator controller asset referenced in the PR summary are not present under Assets/UnityChan in this checkout.
  - The automated demo script scripts/unitychan-demo.sh referenced in the PR summary is not present in this checkout.

## Tests & Verification
- Extensive PlayMode tests added to validate input simulation behaviors and edge cases; FramePressObserver verifies wasPressedThisFrame detection.
- SimulateMouseTests now asserts tool-run completion within 5 seconds to prevent hangs.

## Files Added / Modified (key)
- Added: Assets/Tests/Demo/DemoKeyIndicator.cs
- Added: Assets/Tests/Demo/Editor/DemoKeyboardSceneBuilder.cs
- Added: Assets/Tests/PlayMode/SimulateKeyboardTests.cs
- Added: Assets/UnityChan/Editor/CreateLocatorHere.cs
- Added: .claude/skills/uloop-simulate-keyboard/SKILL.md
- Removed: .claude/skills/simulate-mouse-demo/SKILL.md
- Modified: Assets/Tests/PlayMode/SimulateMouseTests.cs (added timeout)

## Summary of discrepancies to review
- Ensure UnityChanWASDController.cs, scripts/unitychan-demo.sh, and any AnimatorController asset registrations referenced in the PR description are present in the intended branch/commit if they are expected to be part of this change. They were not found in the checked files for this snapshot.

### CODE REVIEW

File: Assets/Tests/Demo/Editor/DemoKeyboardSceneBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+            scaler.matchWidthOrHeight = 0.5f;
+
+            canvasGo.AddComponent<GraphicRaycaster>();
+
+            CreateTitle(canvasGo.transform);
+            CreateKeyboard(canvasGo.transform);
+            CreateStatusText(canvasGo.transform);
+
+            EditorSceneManager.SaveScene(scene, SCENE_PATH);
+            Debug.Log($"[DemoKeyboardSceneBuilder] Scene saved to {SCENE_PATH}");
```
[the remark this answers is not in the document]
Comment (hatayama) [reply Assets/Tests/Demo/Editor/DemoKeyboardSceneBuilder.cs]: Handled in 72f4686c. DemoKeyboardSceneBuilder now captures the SaveScene result and asserts success before logging the saved-scene message, so the success log is no longer emitted on a failed save.

File: Assets/Tests/PlayMode/SimulateKeyboardTests.cs [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+        #endregion
+
+        #region Helpers
+
+        private IEnumerator RunTool(JObject parameters)
+        {
+            Task<BaseToolResponse> task = tool.ExecuteAsync(parameters);
+            yield return new WaitUntil(() => task.IsCompleted);
+            Assert.IsFalse(task.IsFaulted, $"Tool execution should not fault: {task.Exception}");
+            lastResponse = (SimulateKeyboardResponse)task.Result;
```
[the remark this answers is not in the document]
Comment (hatayama) [reply Assets/Tests/PlayMode/SimulateKeyboardTests.cs:516]: Handled in 72f4686c. Added a 5-second timeout guard to RunTool in both SimulateKeyboardTests and SimulateMouseTests, then assert completion before checking task faults, so these PlayMode tests no longer wait indefinitely if ExecuteAsync stalls.

File: .claude/skills/simulate-keyboard-demo/run-demo.sh [hunk 3]
```diff
@@ -0,0 +1,55 @@
+#!/bin/sh
+# SimulateKeyboard demo - exercises all key actions on SimulateKeyboardDemoScene
+
+set -e
```
[the remark this answers is not in the document]
Comment (hatayama) [reply .claude/skills/simulate-keyboard-demo/run-demo.sh]: Handled in 72f4686c. This file is no longer part of the current implementation: the demo shell script was removed and replaced by the new .claude/skills/uloop-simulate-keyboard/SKILL.md workflow, so there is no remaining KeyDown cleanup path to patch in run-demo.sh.

File: .claude/skills/uloop-simulate-keyboard/SKILL.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+```bash
+uloop simulate-keyboard --action <action> --key <key> [options]
+```
+
+### Parameters
+
+| Parameter | Type | Default | Description |
+|-----------|------|---------|-------------|
+| `--action` | enum | `Press` | `Press`, `KeyDown`, `KeyUp` |
+| `--key` | string | (required) | Key name matching Input System Key enum (e.g. `W`, `Space`, `LeftShift`, `A`, `Return`). Case-insensitive. |
```
[the remark this answers is not in the document]
Comment (hatayama) [reply .claude/skills/uloop-simulate-keyboard/SKILL.md]: Addressed in ed811278. Updated the SKILL.md --key example to use Enter so the documentation matches the normalized key naming.

File: Assets/Tests/Demo/Editor/DemoKeyboardSceneBuilder.cs [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+            scaler.matchWidthOrHeight = 0.5f;
+
+            canvasGo.AddComponent<GraphicRaycaster>();
+
+            CreateTitle(canvasGo.transform);
+            CreateKeyboard(canvasGo.transform);
+            CreateStatusText(canvasGo.transform);
+
+            bool saved = EditorSceneManager.SaveScene(scene, SCENE_PATH);
+            Debug.Assert(saved, $"[DemoKeyboardSceneBuilder] Failed to save scene to {SCENE_PATH}");
```
[the remark this answers is not in the document]
Comment (hatayama) [reply Assets/Tests/Demo/Editor/DemoKeyboardSceneBuilder.cs]: Addressed in ed811278. The builder now returns immediately when SaveScene returns false, so the success log only runs after a successful save.

File: Assets/Tests/PlayMode/SimulateKeyboardTests.cs [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+            eventSystemGo.AddComponent<EventSystem>();
+
+            tool = new SimulateKeyboardTool();
+            keyboard = InputSystem.AddDevice<Keyboard>();
+        }
+
+        public override void TearDown()
+        {
+            KeyboardKeyState.Clear();
+            Object.Destroy(eventSystemGo);
```
[the remark this answers is not in the document]
Comment (hatayama) [reply Assets/Tests/PlayMode/SimulateKeyboardTests.cs]: Addressed in ed811278. The PlayMode test teardown now calls KeyboardKeyState.ReleaseAllKeys() so held keys are released from the device state instead of only clearing the in-memory set.

File: Packages/src/Editor/Api/McpTools/SimulateKeyboard/KeyboardKeyState.cs [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+            using (StateEvent.From(keyboard, out InputEventPtr eventPtr))
+            {
+                foreach (Key heldKey in _heldKeys)
+                {
+                    keyboard[heldKey].WriteValueIntoEvent(1f, eventPtr);
+                }
+
+                keyboard[key].WriteValueIntoEvent(pressed ? 1f : 0f, eventPtr);
+                // Updating player state directly avoids editor focus-dependent routing.
+                InputState.Change(keyboard, eventPtr, InputUpdateType.Dynamic);
```
[the remark this answers is not in the document]
Comment (hatayama) [reply Packages/src/Editor/Api/McpTools/SimulateKeyboard/KeyboardKeyState.cs]: Addressed in ed811278. I added KeyboardInputUpdateTypeResolver so simulated key state changes and update waits follow the project's configured Input System update mode instead of hardcoding Dynamic.

File: Assets/Tests/Demo/Editor/DemoKeyboardSceneBuilder.cs [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+            Debug.Log($"[DemoKeyboardSceneBuilder] Scene saved to {SCENE_PATH}");
```
[the remark this answers is not in the document]
Comment (hatayama) [reply Assets/Tests/Demo/Editor/DemoKeyboardSceneBuilder.cs:77]: Addressed in ed811278. A failed SaveScene now triggers an assert and returns before the success log, so the builder no longer reports a successful save on failure.

File: Assets/Tests/PlayMode/SimulateKeyboardTests.cs [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+
+        #endregion
+    }
+
+    public class FramePressObserver : MonoBehaviour
+    {
+        public int SpacePressedFrameCount { get; private set; }
+        public int EnterPressedFrameCount { get; private set; }
+
+        private void Update()
```
[the remark this answers is not in the document]
Comment (hatayama) [reply Assets/Tests/PlayMode/SimulateKeyboardTests.cs]: Addressed in 102707fc. The frame observer no longer relies on MonoBehaviour.Update; it now listens to InputSystem.onAfterUpdate and only samples wasPressedThisFrame for the project's configured Input System update mode.

File: Assets/Tests/PlayMode/SimulateKeyboardTests.cs [hunk 10]
```diff
@@ -0,0 +1,375 @@
...
+        private void OnDisable()
+        {
+            InputSystem.onAfterUpdate -= HandleAfterUpdate;
+        }
+
+        // The tool follows the configured Input System update mode, so the
+        // observer must sample wasPressedThisFrame from the same update loop.
+        private void HandleAfterUpdate()
+        {
+            InputUpdateType expectedUpdateType = ResolveExpectedUpdateType();
```
[the remark this answers is not in the document]
Comment (hatayama) [reply Assets/Tests/PlayMode/SimulateKeyboardTests.cs]: Addressed in 40b057da. The PlayMode test now reuses KeyboardInputUpdateTypeResolver and its IsMatch helper instead of duplicating Input System update-mode resolution inside the observer.

File: Packages/src/package.json [hunk 11]
```diff
@@ -12,6 +12,8 @@
     "url": "https://github.com/hatayama/uLoopMCP"
   },
   "dependencies": {
+    "com.unity.inputsystem": "1.14.2",
+    "com.unity.ugui": "1.0.0",
```
[the remark this answers is not in the document]
Comment (hatayama) [reply Packages/src/package.json:16]: Won't fix. Input System is intentionally listed as a hard dependency — this package requires it for the keyboard simulation overlay (SimulateKeyboardOverlay) and key symbol mapping (KeySymbolMap) which are in the Runtime assembly. The `#if ULOOPMCP_HAS_INPUT_SYSTEM` guards exist only in the Editor tool code to provide a graceful error message, not to make the feature fully optional.

File: Packages/src/Editor/Api/McpTools/SimulateKeyboard/KeyboardKeyState.cs [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+
+namespace io.github.hatayama.uLoopMCP
+{
+    // Tool instances are created fresh per invocation, so held-key state
+    // between KeyDown/KeyUp must be held statically.
+    [InitializeOnLoad]
+    internal static class KeyboardKeyState
+    {
+        private static readonly HashSet<Key> _heldKeys = new();
+        private static readonly HashSet<Key> _transientKeys = new();
```
[the remark this answers is not in the document]
Comment (hatayama) [reply Packages/src/Editor/Api/McpTools/SimulateKeyboard/KeyboardKeyState.cs:16]: Won't fix. Overlapping Press calls for the same key cannot occur in practice — the MCP server processes tool invocations sequentially, and the CLI client waits for each response before sending the next request. A reference-counting approach would add complexity for a scenario that doesn't happen.

File: Assets/Tests/PlayMode/SimulateKeyboardTests.cs [hunk 13]
```diff
@@ -0,0 +1,696 @@
...
+            EnterPressedFrameCount = 0;
+        }
+    }
+
+    public class FrameStateObserver : MonoBehaviour
+    {
+        public int WPressedUpdateCount { get; private set; }
+        public int WReleasedUpdateCount { get; private set; }
+
+        private void Update()
```
[the remark this answers is not in the document]
Comment (hatayama) [reply Assets/Tests/PlayMode/SimulateKeyboardTests.cs:636]: Won't fix. `Keyboard.wKey.isPressed` is persistent state — it stays true from the frame the key is pressed until the frame it's released. Unlike `wasPressedThisFrame`, the observer cannot "miss" it regardless of which update loop the Input System is configured to use. The tool's `WaitForObservationFrames` further guarantees at least 1–2 frames pass before returning, ensuring Update() has executed.

File: Assets/Tests/PlayMode/SimulateKeyboardTests.cs [hunk 14]
```diff
@@ -0,0 +1,696 @@
...
+
+            yield return RunTool(new JObject
+            {
+                ["action"] = KeyboardAction.KeyDown.ToString(),
+                ["key"] = "W"
+            });
+
+            Assert.IsTrue(lastResponse.Success);
+            Assert.IsTrue(KeyboardKeyState.IsKeyHeld(Key.W), "Key should be held after KeyDown");
+            Assert.Greater(frameStateObserver.WPressedUpdateCount, 0, "KeyDown should wait until Update observed the pressed key");
```
[the remark this answers is not in the document]
Comment (hatayama) [reply Assets/Tests/PlayMode/SimulateKeyboardTests.cs:326]: Won't fix. The tool's `ExecuteKeyDown` calls `WaitForObservationFrames` (1–2 frames via `EditorDelay.DelayFrame`) before completing. Since `EditorApplication.update` and `MonoBehaviour.Update()` are synchronized in PlayMode, at least one game-frame Update has run by the time `RunTool` yields back to the test coroutine. The observer checks `isPressed` (persistent), not `wasPressedThisFrame` (transient), so the assertion is reliable.

File: Assets/Tests/PlayMode/SimulateKeyboardTests.cs [hunk 15]
```diff
@@ -0,0 +1,696 @@
...
+
+            SimulateKeyboardSchema parameters = new SimulateKeyboardSchema
+            {
+                Action = KeyboardAction.KeyDown,
+                Key = "W"
+            };
+            CancellationTokenSource cts = new CancellationTokenSource();
+            Task<SimulateKeyboardResponse> task = tool.ExecuteWithCancellationAsync(parameters, cts.Token);
+
+            yield return new WaitUntil(() => KeyboardKeyState.IsKeyHeld(Key.W) || task.IsCompleted);
```
[the remark this answers is not in the document]
Comment (hatayama) [reply Assets/Tests/PlayMode/SimulateKeyboardTests.cs:487]: Won't fix. Unity's test runner already enforces a default timeout at the test level. Adding an inline timeout to every `WaitUntil` adds noise without meaningful safety — if the key state is never applied, the test-level timeout will catch it.

File: Assets/Tests/Demo/Editor/DemoKeyboardSceneBuilder.cs [hunk 16]
```diff
@@ -0,0 +1,284 @@
...
+                new KeyDef(Key.B, "B"), new KeyDef(Key.N, "N"),
+                new KeyDef(Key.M, "M"),
+            };
+            BuildRow(panel.transform, row4, row4Y);
+
+            // Row 5: LCtrl LAlt LMeta Space RMeta RAlt RCtrl
+            float row5Y = row4Y - (KEY_SIZE + ROW_GAP);
+            string metaLabel = IsMac ? "\u2318" : "Win";
+            string altLabel = IsMac ? "\u2325" : "Alt";
+            string ctrlLabel = IsMac ? "\u2303" : "Ctrl";
```
[the remark this answers is not in the document]
Comment (hatayama) [reply Assets/Tests/Demo/Editor/DemoKeyboardSceneBuilder.cs:203]: Won't fix. `DemoKeyIndicator.Awake()` already resolves labels at runtime via `KeySymbolMap.GetSymbol()`, overriding whatever was serialized at scene-build time. The scene-build labels are just placeholders — they never appear to the user.

File: scripts/test-meta-key-overlay.sh [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+screenshot
+release LeftMeta
+release LeftAlt
+release LeftCtrl
+sleep 1
+
+# --- Special keys (Press) ---
+
+echo "[7/7] Special keys: Tab(⇥) Esc Enter(⏎) Backspace(⌫) Space(␣) Arrows(↑↓←→)"
+press Tab 0.3
```
[the remark this answers is not in the document]
Comment (hatayama) [reply scripts/test-meta-key-overlay.sh:94]: Won't fix. Step 7 is designed for visual verification while the script runs — the overlay briefly appears for each special key. Screenshots are only captured for modifier keys (Steps 1–6) where persistent hold state makes them useful. Press-based keys fade within 0.7s, making screenshot timing unreliable.

File: scripts/test-unitychan-keyboard-input.ps1 [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+    Press -Key Space -Duration 0.3
+    Start-Sleep -Seconds 2.0
+    Release -Key W
+    HoldFor -Duration 0.8 -Keys RightArrow
+    HoldFor -Duration 0.8 -Keys DownArrow
+    HoldFor -Duration 0.8 -Keys LeftArrow
+    Start-Sleep -Seconds 0.3
+
+    # Final idle jump
+    Press -Key Enter -Duration 0.1
```
[the remark this answers is not in the document]
Comment (hatayama) [reply scripts/test-unitychan-keyboard-input.ps1:84]: Won't fix. This is a 1:1 port of the shell script. The "Final idle jump" section tests rapid key input and overlay rendering, not UnityChan's jump action. The Enter key was chosen intentionally to exercise a different key than Space.

File: Assets/Tests/Demo/DemoKeyIndicator.cs [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+
+            // Scene file bakes modifier labels for one platform; fix at runtime
+            string rawName = targetKey.ToString();
+            string symbol = KeySymbolMap.GetSymbol(rawName);
+            if (rawName != symbol)
+            {
+                Text text = GetComponentInChildren<Text>();
+                if (text != null)
+                {
+                    text.text = symbol;
```
[the remark this answers is not in the document]
Comment (hatayama) [reply Assets/Tests/Demo/DemoKeyIndicator.cs:30]: Won't fix. The demo keyboard keys use fixed-size squares with centered text. The current glyph rendering looks fine at the default font size — no user complaints. The overlay (`SimulateKeyboardOverlay`) already handles glyph sizing with `GLYPH_FONT_SIZE`, but the demo scene keys have different layout constraints.
